local_pathfinding/objectives.py

```diff
@@ -24,66 +35,308 @@ def motionCost(self, s1: ob.SE2StateSpace, s2: ob.SE2StateSpace) -> ob.Cost:
+        theta_max = wind_direction + math.radians(UPWIND_MAX_ANGLE_DEGREES)
+
+        return WindObjective.is_angle_between(theta_min, boat_direction, theta_max)
 
     @staticmethod
-    def _is_downwind(wind_direction_radians: float, boat_direction_radians: float) -> bool:
-        raise NotImplementedError
+    def is_downwind(wind_direction: float, boat_direction: float) -> bool:
+        """Generates the cost associated with the downwind direction
+
+        Args:
+            wind_direction (float): The true wind direction (radians). [-pi, pi)
+            boat_direction_radians (float)): The direction of the boat (radians). [-pi, pi)
+
+        Returns:
+            bool: The cost associated with the downwind direction
+        """
+        downwind_wind_direction = (wind_direction + math.pi) % (2 * math.pi)
+
+        theta_min = downwind_wind_direction - math.radians(DOWNWIND_MAX_ANGLE_DEGREES)
+
+        theta_max = downwind_wind_direction + math.radians(DOWNWIND_MAX_ANGLE_DEGREES)
+
+        return WindObjective.is_angle_between(theta_min, boat_direction, theta_max)
+
+    @staticmethod
+    def is_angle_between(first_angle: float, middle_angle: float, second_angle: float) -> bool:
+        """Determines whether an angle is between two other angles.
+
+        Args:
+            first_angle (float): The first bounding angle in radians.
+            middle_angle (float): The angle in question in radians.
+            second_angle (float): The second bounding angle in radians.
+
+        Returns:
+            bool: True when `middle_angle` is not in the reflex angle of
+                `first_angle` and `second_angle`, false otherwise.
+        """
+        if first_angle < second_angle:
+            if (
+                second_angle - math.pi == first_angle
+            ):  # Assume all angles are between first and second when pi degrees apart
+                return middle_angle != first_angle and middle_angle != second_angle
+            elif second_angle - math.pi < first_angle:
+                return middle_angle > first_angle and middle_angle < second_angle
+            else:
+                return middle_angle < first_angle or middle_angle > second_angle
+        else:
+            return WindObjective.is_angle_between(second_angle, middle_angle, first_angle)
 
 
 def get_sailing_objective(
     space_information, simple_setup, heading_degrees: float, wind_direction_degrees: float
 ) -> ob.OptimizationObjective:
     objective = ob.MultiOptimizationObjective(si=space_information)
-    objective.addObjective(objective=DistanceObjective(space_information), weight=1.0)
+    objective.addObjective(objective=DistanceObjective(space_information, "latlon"), weight=1.0)
     objective.addObjective(
-        objective=MinimumTurningObjective(space_information, simple_setup, heading_degrees),
+        objective=MinimumTurningObjective(
+            space_information, simple_setup, heading_degrees, "goal_heading"
+        ),
```

Comment:
Any reason why they override the default `implementation`s?

Reply:
Just because.....